### ISSUE TITLE
[product icon themes] Finalize icon contribution API

### ISSUE THREAD

**Issue by aeschli** [microsoft/vscode#119101]
The icon contribution point allow extensions to define a new icon by id, along with a default icon.
That icon id can then be used by the extension (or any other extensions that depend on the extension) at the places where [ThemeIcon]([ThemeIcon](https://github.com/microsoft/vscode/blob/e88863d191fed9c81a0087ff225ef7daf79c40b8/src/vs/vscode.d.ts#L800)) can be used `new ThemeIcon("iconId")`, [markdown strings](https://github.com/microsoft/vscode/blob/e88863d191fed9c81a0087ff225ef7daf79c40b8/src/vs/vscode.d.ts#L2516) (`$(iconId)`) and as icons in certain contribution points.

Product icon themes can redefine the icon (if they know about the icon id)

```
"contributes": {
	"icons": [
		{
			"id": "distro-ubuntu",
			"description": "Ubuntu icon",
			"default": {
				"fontId": "distro-icon-font",
				"fontCharacter": "\\E001"
			}
		},
		{
			"id": "distro-fedora",
			"description": "Fedora icon",
			"default": {
				"fontId": "distro-icon-font",
				"fontCharacter": "\\E002"
			}
		}
	],
	"iconFonts": [
		{
			"id": "distro-icon-font",
			"src": [
				{
					"path": "./distroicons.woff",
					"format": "woff"
				}
			]
		}
	]
}

````
In this example, an extension defines two new icon ids `distro-ubuntu` and `distro-fedora` along with default icon definitions. The icons are defined in an icon font at the given font character. The icon font is defined in  a new contribution point `iconFonts`.

## Update

See https://github.com/microsoft/vscode/issues/119101#issuecomment-1032860411 for the latest proposal that is implemented in the latest 1.65 insiders builds
```
	"icons": {
		"distro-ubuntu": {
			"description": "Ubuntu icon",
			"default": {
				"fontPath": "./distroicons.woff",
				"fontCharacter": "\\E001"
			}
		},
		"distro-fedora": {
			"description": "Ubuntu icon",
			"default": {
				"fontPath": "./distroicons.woff",
				"fontCharacter": "\\E002"
			}
		}
	}
```

**Comment by aeschli** [comment]
Planed for April. I still want to measure if having many fonts can become a performance issue.

**Comment by matthewjamesadam** [comment]
@aeschli I am very interested in this proposal.  One use case that is compelling to me is being able to use an extended set of icons in the status bar, specifically so that an extension could use icons in StatusBarItems that are domain-specific, that would likely never be added as octicons.

Has there been any progress or discussion about adopting this proposal?  I would be willing to put work towards this feature, if there's any way I can help.

**Comment by aeschli** [comment]
In the API call we discussed that it's not nice to have a separate "iconFonts" section which is internal to the extension.
Looking at this I think we can condense it. The icon path is enough, and we can require it to end with `woff` or `tiff`. There are currently no other font properties supported. If icon size becomes a requirement, we can add it to each icon.
```json
	"icons": [
		{
			"id": "distro-ubuntu",
			"description": "Ubuntu icon",
			"default": {
				"fontPath": "./distroicons.woff"
				"fontCharacter": "\\E001"
			}
		},
		{
			"id": "distro-fedora",
			"description": "Fedora icon",
			"default": {
				"fontPath": "./distroicons.woff",
				"fontCharacter": "\\E002"
			}
		}
	]
```
@jrieken @mjbvz ?

**Comment by misolori** [comment]
This broke the GitHub Copilot extension for a couple of days (see https://github.com/github/feedback/discussions/11684) so suggested they update their `icons` contribution to:

```json
		"icons": {
			"copilot-logo": {
				"description": "GitHub Copilot Icon",
				"default": {
					"fontPath": "assets/copilot.woff",
					"fontCharacter": "\\0041"
				}
			},
			"copilot-warning": {
				"description": "GitHub Copilot Warning",
				"default": {
					"fontPath": "assets/copilot.woff",
					"fontCharacter": "\\0042"
				}
			},
			"copilot-notconnected": {
				"description": "GitHub Copilot Not Connected",
				"default": {
					"fontPath": "assets/copilot.woff",
					"fontCharacter": "\\0043"
				}
			}
		},
```

@aeschli is there no longer a shorthand way to use `iconFonts` when defined? Feels a bit silly to have to specify the font path for every contributed icon (if you've got a lot this can be tedious)

**Comment by aeschli** [comment]
@misolori I filed A PR for copilot a week back (https://github.com/github/copilot/issues/1759)

The outcome of the API call was to remove the `iconFonts` contribution point as the contributed icon fonts were internal to the extension and more like a helper. No other contribution point is like that. Given that all that it describes was the path to the font and the format of the font can be deducted from the file name, we went with the font path. Maybe don't use the most complicated font path...
